pkg/apis/serving/k8s_validation.go

```diff
@@ -257,7 +260,21 @@ func ValidatePodSpec(ps corev1.PodSpec) *apis.FieldError {
 		errs = errs.Also(ValidateContainer(ps.Containers[0], volumes).
 			ViaFieldIndex("containers", 0))
 	default:
-		errs = errs.Also(apis.ErrMultipleOneOf("containers"))
+		cfg := config.FromContextOrDefaults(ctx).Defaults
+		if !cfg.EnableMultiContainer {
+			errs = errs.Also(apis.ErrMultipleOneOf("containers"))
+		} else {
+			errs = errs.Also(ValidateMultiContainerPorts(ps.Containers)).ViaField("containers")
+			for i := range ps.Containers {
+				// probes are not allowed other than serving container
+				// ref: https://docs.google.com/document/d/1XjIRnOGaq9UGllkZgYXQHuTQmhbECNAOk6TT6RNfJMw/edit?disco=AAAAEHNSwZU
+				if len(ps.Containers[i].Ports) == 0 {
+					errs = errs.Also(ValidateSidecarContainer(ps.Containers[i], volumes).ViaFieldIndex("containers", i))
+				} else {
+					errs = errs.Also(ValidateContainer(ps.Containers[i], volumes).ViaFieldIndex("containers", i))
+				}
+			}
+		}
```

Comment:
Should we pull the whole thingy (from `switch len(ps.Containers)`) to `validateContainers` function? It would simplify the switch flow as well.